File: ports/wasm-dist/sync/sync.mjs
/**
 * sync.mjs — frameless memory synchronization service
 *
 * Provides named regions (slab, ring, slots) backed by a tiny WASM binary.
 * Auto-selects transport: SharedArrayBuffer for zero-copy when available,
 * postMessage + Transferable ArrayBuffer when not.
 *
 * Usage:
 *   const bus = await SyncBus.create();
 *   bus.region('serial_tx', { size: 4096, type: 'ring' });
 *   bus.region('gpio', { size: 384, type: 'slots', slotSize: 12 });
 *   const port = bus.port();
 *   port.push('serial_tx', new Uint8Array([72, 105]));
 *   const bytes = port.drain('serial_tx');
 */

const TYPE_SLAB  = 0;
const TYPE_RING  = 1;
const TYPE_SLOTS = 2;

// ── SyncBus ──
// Owns the WASM instance and memory. Creates ports (endpoints).

export class SyncBus {
    constructor() {
        this._instance = null;
        this._memory = null;
        this._exports = null;
        this._regions = new Map();  // name → { index, hash, type }
        this._shared = false;
    }

    /**
     * Create a bus backed by a sync.wasm binary.
     * @param {object} [options]
     * @param {string} [options.wasmUrl] — URL to sync.wasm (default: same dir)
     * @param {boolean} [options.shared] — force SharedArrayBuffer (default: auto-detect)
     */
    static async create(options = {}) {
        const bus = new SyncBus();
        await bus._init(options);
        return bus;
    }

    async _init(options) {
        const canShare = typeof SharedArrayBuffer !== 'undefined'
            && typeof Atomics !== 'undefined';
        this._shared = options.shared ?? canShare;

        const memoryDescriptor = {
            initial: 10,    // 640KB — bus struct + 512KB arena
            maximum: 16,    // grow up to 1MB if needed
        };
        if (this._shared) memoryDescriptor.shared = true;

        this._memory = new WebAssembly.Memory(memoryDescriptor);

        const wasmUrl = options.wasmUrl
            ?? new URL('./sync.wasm', import.meta.url).href;
        const wasmBytes = await fetch(wasmUrl).then(r => r.arrayBuffer());
        const module = await WebAssembly.compile(wasmBytes);
        this._instance = await WebAssembly.instantiate(module, {
            env: { memory: this._memory },
        });
        this._exports = this._instance.exports;
    }

    /** The raw memory buffer. Shared if SAB available. */
    get buffer() { return this._memory.buffer; }

    /** Whether the bus uses SharedArrayBuffer. */
    get isShared() { return this._shared; }

    /**
     * Define a named region.
     * @param {string} name
     * @param {object} opts
     * @param {number} opts.size — data size in bytes
     * @param {'slab'|'ring'|'slots'} [opts.type='slab']
     * @param {number} [opts.slotSize] — required for type='slots'
     * @returns {number} data offset in linear memory
     */
    region(name, opts) {
        const { size, type = 'slab', slotSize = 0 } = opts;
        const typeCode = { slab: TYPE_SLAB, ring: TYPE_RING, slots: TYPE_SLOTS }[type];
        if (typeCode === undefined) throw new Error(`Unknown region type: ${type}`);

        // Write name into WASM memory for hashing
        const nameBytes = new TextEncoder().encode(name);
        const namePtr = this._writeTmp(nameBytes);
        const offset = this._exports.region_create(namePtr, nameBytes.length,
            size, typeCode, slotSize);
        if (offset === 0) throw new Error(`Failed to create region '${name}'`);

        const hash = this._exports.hash_name(namePtr, nameBytes.length);
        const index = this._exports.region_find(hash);
        this._regions.set(name, { index, hash, type: typeCode, offset, size, slotSize });
        return offset;
    }

    /** Create an endpoint port for this bus. */
    port() {
        return new SyncPort(this);
    }

    /**
     * Create a transferable descriptor for sending to a Worker.
     * The Worker calls SyncBus.fromTransfer(desc) to reconstruct.
     */
    transfer() {
        const regions = {};
        for (const [name, info] of this._regions) {
            regions[name] = info;
        }
        return {
            memory: this._memory,
            regions,
            shared: this._shared,
        };
    }

    /**
     * Reconstruct a bus from a transferred descriptor (Worker side).
     * No WASM needed — the Worker reads/writes memory directly.
     */
    static fromTransfer(desc) {
        const bus = new SyncBus();
        bus._memory = desc.memory;
        bus._shared = desc.shared;
        bus._exports = null;  // Worker doesn't need WASM exports for read/write
        for (const [name, info] of Object.entries(desc.regions)) {
            bus._regions.set(name, info);
        }
        return bus;
    }

    // Write bytes to a scratch area for passing strings to WASM.
    // Uses the last 256 bytes of the first memory page.
    _writeTmp(bytes) {
        const tmpOffset = 65536 - 256;
        new Uint8Array(this._memory.buffer).set(bytes, tmpOffset);
        return tmpOffset;
    }

    _regionInfo(name) {
        const info = this._regions.get(name);
        if (!info) throw new Error(`Unknown region: ${name}`);
        return info;
    }
}


// ── SyncPort ──
// An endpoint view over the bus. Reads and writes named regions.
// Tracks local sequence numbers for dirty detection.

export class SyncPort {
    constructor(bus) {
        this._bus = bus;
        this._lastSeq = new Map();  // name → last seen seq
    }

    // ── Dirty detection ──

    /** Has this region changed since we last checked? */
    changed(name) {
        const info = this._bus._regionInfo(name);
        const seq = this._readSeq(info);
        if (!this._lastSeq.has(name)) {
            this._lastSeq.set(name, seq);  // first check: treat as clean
            return false;
        }
        return seq !== this._lastSeq.get(name);
    }

    /** Mark region as "seen" at current sequence. */
    acknowledge(name) {
        const info = this._bus._regionInfo(name);
        this._lastSeq.set(name, this._readSeq(info));
    }

    // ── Slab operations ──

    /** Read entire slab as a typed array view (zero-copy if SAB). */
    read(name) {
        const info = this._bus._regionInfo(name);
        this._lastSeq.set(name, this._readSeq(info));
        return new Uint8Array(this._bus.buffer, info.offset, info.size);
    }

    /** Read slab as Uint16Array (e.g., RGB565 framebuffer). */
    read16(name) {
        const info = this._bus._regionInfo(name);
        this._lastSeq.set(name, this._readSeq(info));
        return new Uint16Array(this._bus.buffer, info.offset, info.size / 2);
    }

    /** Write to slab at byte offset. Marks dirty. */
    write(name, offset, data) {
        const info = this._bus._regionInfo(name);
        new Uint8Array(this._bus.buffer).set(data, info.offset + offset);
        this._markDirty(info);
    }

    /** Write entire slab. Marks dirty. */
    writeSlab(name, data) {
        this.write(name, 0, data);
    }

    // ── Ring operations ──

    /** Push bytes into a ring buffer. Returns bytes written. */
    push(name, bytes) {
        const info = this._bus._regionInfo(name);
        const buf = this._bus.buffer;
        const readHeadPtr = info.offset;      // u32 read_head at start
        const dataBase = info.offset + 4;
        const cap = info.size;
        const view = new DataView(buf);
        const data = new Uint8Array(buf);

        let wh = this._readAux(info);
        const rh = view.getUint32(readHeadPtr, true);
        let written = 0;

        for (let i = 0; i < bytes.length; i++) {
            const next = (wh + 1) % cap;
            if (next === rh) break;  // full
            data[dataBase + wh] = bytes[i];
            wh = next;
            written++;
        }
        if (written > 0) {
            this._writeAux(info, wh);
            this._bumpSeq(info);
        }
        return written;
    }

    /** Drain all available bytes from a ring buffer. */
    drain(name) {
        const info = this._bus._regionInfo(name);
        const buf = this._bus.buffer;
        const readHeadPtr = info.offset;
        const dataBase = info.offset + 4;
        const cap = info.size;
        const view = new DataView(buf);
        const data = new Uint8Array(buf);

        const wh = this._readAux(info);
        let rh = view.getUint32(readHeadPtr, true);

        if (rh === wh) return new Uint8Array(0);

        const result = [];
        while (rh !== wh) {
            result.push(data[dataBase + rh]);
            rh = (rh + 1) % cap;
        }
        view.setUint32(readHeadPtr, rh, true);
        return new Uint8Array(result);
    }

    /** Bytes available to read from ring. */
    available(name) {
        const info = this._bus._regionInfo(name);
        const view = new DataView(this._bus.buffer);
        const rh = view.getUint32(info.offset, true);
        const wh = this._readAux(info);
        return (wh - rh + info.size) % info.size;
    }

    // ── Slot operations ──

    /** Read a single slot as Uint8Array. */
    readSlot(name, slotIndex) {
        const info = this._bus._regionInfo(name);
        const off = info.offset + slotIndex * info.slotSize;
        return new Uint8Array(this._bus.buffer, off, info.slotSize);
    }

    /** Write a slot. Marks slot dirty. */
    writeSlot(name, slotIndex, data) {
        const info = this._bus._regionInfo(name);
        const off = info.offset + slotIndex * info.slotSize;
        new Uint8Array(this._bus.buffer).set(data, off);
        // Set dirty bit
        const view = new DataView(this._bus.buffer);
        const entryBase = 16 + info.index * 32;
        const dirty = view.getUint32(entryBase + 20, true);
        view.setUint32(entryBase + 20, dirty | (1 << slotIndex), true);
        this._bumpSeq(info);
    }

    /** Get dirty slot bitmask and clear it. Returns array of dirty slot indices. */
    dirtySlots(name) {
        const info = this._bus._regionInfo(name);
        const view = new DataView(this._bus.buffer);
        const entryBase = 16 + info.index * 32;
        const mask = view.getUint32(entryBase + 20, true);
        view.setUint32(entryBase + 20, 0, true);
        const slots = [];
        for (let i = 0; i < 32; i++) {
            if (mask & (1 << i)) slots.push(i);
        }
        return slots;
    }

    // ── Global ──

    /** Has anything on the bus changed since last check? */
    get globalChanged() {
        const view = new DataView(this._bus.buffer);
        const seq = view.getUint32(8, true);
        const last = this._lastGlobalSeq ?? -1;
        this._lastGlobalSeq = seq;
        return seq !== last;
    }

    // ── Internal helpers ──

    _readSeq(info) {
        const view = new DataView(this._bus.buffer);
        return view.getUint32(16 + info.index * 32 + 16, true);
    }

    _bumpSeq(info) {
        const view = new DataView(this._bus.buffer);
        const entryBase = 16 + info.index * 32;
        const seq = view.getUint32(entryBase + 16, true) + 1;
        view.setUint32(entryBase + 16, seq, true);
        // Bump global seq
        const gseq = view.getUint32(8, true) + 1;
        view.setUint32(8, gseq, true);
    }

    _markDirty(info) {
        const view = new DataView(this._bus.buffer);
        const entryBase = 16 + info.index * 32;
        view.setUint32(entryBase + 20, 1, true);
        this._bumpSeq(info);
    }

    _readAux(info) {
        const view = new DataView(this._bus.buffer);
        return view.getUint32(16 + info.index * 32 + 28, true);
    }

    _writeAux(info, val) {
        const view = new DataView(this._bus.buffer);
        view.setUint32(16 + info.index * 32 + 28, val, true);
    }
}

File: ports/wasm-dist/sync/sync.c
// sync.c — frameless memory synchronization service
//
// Compiles to a standalone WASM binary (~2KB). No libc, no WASI.
// All storage is in linear memory. JS reads/writes directly.
//
// Build:
//   clang --target=wasm32 -nostdlib -O2 \
//     -Wl,--no-entry -Wl,--export-dynamic \
//     -Wl,--import-memory \
//     -o sync.wasm sync.c

#include <stdint.h>

// ── Memory layout ──
//
// A global struct at a known address in linear memory.
// JS discovers addresses via exported query functions.

#define MAX_REGIONS  64

// Region types
enum { TYPE_SLAB = 0, TYPE_RING = 1, TYPE_SLOTS = 2 };

// Region table entry (32 bytes)
typedef struct {
    uint32_t name_hash;
    uint32_t data_offset;
    uint32_t data_size;
    uint32_t region_type;   // 0=slab, 1=ring, 2=slots
    uint32_t seq;
    uint32_t dirty;
    uint32_t slot_size;
    uint32_t aux;           // ring: write_head; slots: slot_count
} region_entry_t;

// Bus state — lives in the data segment, linker places it
static struct {
    uint32_t        region_count;
    uint32_t        total_allocated;
    uint32_t        global_seq;
    uint32_t        reserved;
    region_entry_t  entries[MAX_REGIONS];
} bus;

// Data regions are allocated starting after the bus struct.
// We reserve a large arena for dynamic allocation.
#define ARENA_SIZE (512 * 1024)
static uint8_t arena[ARENA_SIZE];

// ── Helpers ──

static uint32_t fnv1a(const uint8_t *s, int len) {
    uint32_t h = 2166136261u;
    for (int i = 0; i < len; i++) {
        h ^= s[i];
        h *= 16777619u;
    }
    return h;
}

static int has_region(uint32_t hash) {
    for (uint32_t i = 0; i < bus.region_count; i++) {
        if (bus.entries[i].name_hash == hash) return 1;
    }
    return 0;
}

// ── Exported functions ──

// Create a named region. Returns data offset in linear memory, or 0.
__attribute__((export_name("region_create")))
uint32_t region_create(const uint8_t *name, int name_len,
                       uint32_t size, uint32_t type, uint32_t slot_size) {
    if (bus.region_count >= MAX_REGIONS) return 0;

    uint32_t hash = fnv1a(name, name_len);
    if (has_region(hash)) return 0;

    // Align to 8 bytes
    uint32_t arena_offset = (bus.total_allocated + 7) & ~7u;

    // Rings need 4 extra bytes for read_head
    uint32_t total = (type == TYPE_RING) ? size + 4 : size;
    if (arena_offset + total > ARENA_SIZE) return 0;

    // Compute absolute address in linear memory
    uint32_t abs_offset = (uint32_t)(uintptr_t)&arena[arena_offset];

    region_entry_t *r = &bus.entries[bus.region_count];
    r->name_hash   = hash;
    r->data_offset = abs_offset;
    r->data_size   = size;
    r->region_type = type;
    r->seq         = 0;
    r->dirty       = 0;
    r->slot_size   = slot_size;
    r->aux         = 0;

    if (type == TYPE_SLOTS && slot_size > 0) {
        r->aux = size / slot_size;
    }

    // Zero the region
    for (uint32_t i = 0; i < total; i++) arena[arena_offset + i] = 0;

    bus.total_allocated = arena_offset + total;
    bus.region_count++;

    return abs_offset;
}

__attribute__((export_name("region_find")))
int32_t region_find(uint32_t name_hash) {
    for (uint32_t i = 0; i < bus.region_count; i++) {
        if (bus.entries[i].name_hash == name_hash) return (int32_t)i;
    }
    return -1;
}

__attribute__((export_name("hash_name")))
uint32_t hash_name(const uint8_t *name, int len) {
    return fnv1a(name, len);
}

// ── Slab operations ──

__attribute__((export_name("slab_mark_dirty")))
void slab_mark_dirty(uint32_t idx) {
    bus.entries[idx].dirty = 1;
    bus.entries[idx].seq++;
    bus.global_seq++;
}

__attribute__((export_name("slab_clear_dirty")))
uint32_t slab_clear_dirty(uint32_t idx) {
    bus.entries[idx].dirty = 0;
    return bus.entries[idx].seq;
}

// ── Ring operations ──
// Layout at data_offset: [read_head:u32] [data: data_size bytes]
// write_head lives in entry.aux.

__attribute__((export_name("ring_push")))
uint32_t ring_push(uint32_t idx, const uint8_t *src, uint32_t len) {
    region_entry_t *r = &bus.entries[idx];
    uint32_t *rh_ptr = (uint32_t *)(uintptr_t)r->data_offset;
    uint8_t  *data   = (uint8_t *)(uintptr_t)(r->data_offset + 4);
    uint32_t cap = r->data_size;
    uint32_t wh = r->aux;
    uint32_t rh = *rh_ptr;

    uint32_t written = 0;
    for (uint32_t i = 0; i < len; i++) {
        uint32_t next = (wh + 1) % cap;
        if (next == rh) break;
        data[wh] = src[i];
        wh = next;
        written++;
    }
    r->aux = wh;
    if (written > 0) { r->seq++; bus.global_seq++; }
    return written;
}

__attribute__((export_name("ring_drain")))
uint32_t ring_drain(uint32_t idx, uint8_t *dst, uint32_t max_len) {
    region_entry_t *r = &bus.entries[idx];
    uint32_t *rh_ptr = (uint32_t *)(uintptr_t)r->data_offset;
    uint8_t  *data   = (uint8_t *)(uintptr_t)(r->data_offset + 4);
    uint32_t cap = r->data_size;
    uint32_t wh = r->aux;
    uint32_t rh = *rh_ptr;

    uint32_t count = 0;
    while (rh != wh && count < max_len) {
        dst[count++] = data[rh];
        rh = (rh + 1) % cap;
    }
    *rh_ptr = rh;
    return count;
}

__attribute__((export_name("ring_available")))
uint32_t ring_available(uint32_t idx) {
    region_entry_t *r = &bus.entries[idx];
    uint32_t rh = *(uint32_t *)(uintptr_t)r->data_offset;
    uint32_t wh = r->aux;
    return (wh - rh + r->data_size) % r->data_size;
}

// ── Slot operations ──

__attribute__((export_name("slot_mark_dirty")))
void slot_mark_dirty(uint32_t idx, uint32_t slot) {
    if (slot < 32) bus.entries[idx].dirty |= (1u << slot);
    bus.entries[idx].seq++;
    bus.global_seq++;
}

__attribute__((export_name("slot_drain_dirty")))
uint32_t slot_drain_dirty(uint32_t idx) {
    uint32_t mask = bus.entries[idx].dirty;
    bus.entries[idx].dirty = 0;
    return mask;
}

// ── Query functions ──

__attribute__((export_name("bus_addr")))
uint32_t bus_addr(void) { return (uint32_t)(uintptr_t)&bus; }

__attribute__((export_name("region_count")))
uint32_t region_count(void) { return bus.region_count; }

__attribute__((export_name("region_seq")))
uint32_t region_seq(uint32_t i) { return bus.entries[i].seq; }

__attribute__((export_name("region_offset")))
uint32_t region_offset(uint32_t i) { return bus.entries[i].data_offset; }

__attribute__((export_name("region_size")))
uint32_t region_size(uint32_t i) { return bus.entries[i].data_size; }

__attribute__((export_name("region_type")))
uint32_t region_type(uint32_t i) { return bus.entries[i].region_type; }

__attribute__((export_name("region_dirty")))
uint32_t region_dirty(uint32_t i) { return bus.entries[i].dirty; }

__attribute__((export_name("global_seq")))
uint32_t global_seq(void) { return bus.global_seq; }
